### PR TITLE
Add emperor endpoint implementation

### PR DESCRIPTION
**Pull request by gwarmstrong**
Adds an implementation of the proposed Emperor endpoint. Including:

* Added a mechanism for loading Q2 PCoA artifacts via `resources`
* Added `metadata_categories` parameter to the REST specification
* A `PCoA` repo for getting the PCoA data
* An `implementation` for the endpoint, which grabs the data and formats it appropriately

### CODE REVIEW

File: microsetta_public_api/api/emperor.py [hunk 1]
```diff
@@ -1,2 +1,43 @@
...
+    # grab the sample ids from the PCoA
+    samples = pcoa.samples.index
+
+    # metadata for samples not in the repo will be filled in as None
+    metadata = metadata_repo.get_metadata(metadata_categories,
+                                          sample_ids=samples)
+
+    response = dict()
+    response['decomposition'] = {
+        "coordinates": pcoa.samples.values.tolist(),
```
Comment (wasade) [inline microsetta_public_api/api/emperor.py:36]: to reduce what goes out, would it be feasible to limit this to the first 3 axes?

File: microsetta_public_api/api/emperor.py [hunk 2]
```diff
@@ -1,2 +1,43 @@
...
+        "percents_explained": list(100 * prop for
+                                   prop in pcoa.proportion_explained),
+        "sample_ids": list(samples),
+    }
+    response["metadata"] = metadata.values.tolist()
```
Comment (wasade) [inline microsetta_public_api/api/emperor.py:41]: is it necessary that the row order of `metadata` is consistent with the row order of `pcoa.samples`?

File: microsetta_public_api/resources.py [hunk 3]
```diff
@@ -22,6 +24,22 @@ def _dict_of_paths_to_alpha_data(dict_of_qza_paths, resource_name):
...
+def _dict_of_paths_to_pcoa(dict_of_qza_paths, resource_name):
+    _validate_dict_of_paths(dict_of_qza_paths,
+                            resource_name)
+    new_resource = _replace_paths_with_qza(dict_of_qza_paths,
+                                           PCoAResults,
+                                           view_type=OrdinationResults)
+    return new_resource
+
+
+def _dict_of_dict_of_paths_to_pcoa(dict_of_dict_of_qza_paths, resource_name):
```
Comment (wasade) [inline microsetta_public_api/resources.py:36]: A refactor for the future would be to create some objects to represent this complex data type

File: microsetta_public_api/api/emperor.py (hunk 1 above)
Comment (gwarmstrong) [reply microsetta_public_api/api/emperor.py:36]: Similar to metadata, shouldn't the onus for that fall on whoever is configuring the server?

Comment (wasade) [reply microsetta_public_api/api/emperor.py:36]: fair

File: microsetta_public_api/api/emperor.py (hunk 2 above)
Comment (gwarmstrong) [reply microsetta_public_api/api/emperor.py:41]: Good catch

File: microsetta_public_api/repo/_metadata_repo.py [hunk 4]
```diff
@@ -70,6 +74,28 @@ def category_values(self, category, exclude_na=True):
...
+        if isinstance(sample_id, str):
+            return sample_id in self._metadata.index
+        else:
+            index = set(self._metadata.index)
+            return [id_ in index for id_ in sample_id]
+
+    def get_metadata(self, categories, sample_ids=None):
+        md = self._metadata[categories]
+        if sample_ids is not None:
+            md = md.reindex(sample_ids, fill_value=None)
```
Comment (gwarmstrong) [inline microsetta_public_api/repo/_metadata_repo.py:94]: @wasade the metadata is ordered based on samples here

File: microsetta_public_api/api/emperor.py (hunk 2 above)
Comment (gwarmstrong) [reply microsetta_public_api/api/emperor.py:41]: I actually already handled that in the call to `metadata_repo.get_metadata`

File: microsetta_public_api/resources.py (hunk 3 above)
Comment (gwarmstrong) [reply microsetta_public_api/resources.py:36]: Yes. For the milestone item "update results API to provide shotgun results in addition to amplicon" I intend to refactor this whole module so it is much more OOP

File: microsetta_public_api/api/emperor.py (hunk 2 above)
Comment (wasade) [reply microsetta_public_api/api/emperor.py:41]: oh awesome

**Pull request merged**